### PR TITLE
Update code contribution guidelines.

### PR DESCRIPTION
**Pull request by patrickvacek**
Removed some redundant/obvious/irrelevant bits and added some more detail to reflect our current standards and preferences.

This is still up for discussion, so if you disagree with any of this or think something is missing, please speak up.

### CODE REVIEW

**Comment by patrickvacek** [comment]
Although I appreciate the quick approval, I'm going to wait until some more of our team is back from vacation before merging, as I'd like to get a bit wider input on this.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -13,30 +13,28 @@ Code quality and style
...
+   - Developers should not create large, month-long, multiple-feature branches.
+     - The larger the PR, the harder it is to review, the more likely it will need to be rebased, and the harder it is to rebase it.
+     - If a PR changes thousands of lines of code, please consider splitting it into smaller PRs.
+   - Developers should aim to have their code merged at least once every week.
+   - Multiple small fixes or refactorings can be grouped together in one PR, but each independent fix or refactoring should be a unique commit.
+* All code must be reviewed before it is merged to the master branch.
+   - The reviewer can not be the code author.
+   - All the reviewer's concerns must be resolved to reviewer's satisfaction.
+   - Reviews should enforce the coding guidelines for the project.
+* Bugs should be reproduced with a failing test case before they are resolved.
```
Comment (xaviergully) [inline CONTRIBUTING.md]: 👍

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -13,30 +13,28 @@ Code quality and style
...
+* Only working code that passes all tests in CI will be merged into the master branch.
+   - The master branch should always be in a deployable state.
+   - Functionally incomplete code is not necessarily undeployable.
+* Pull requests should only contain a single feature, fix, or refactoring.
+   - Developers should not create large, month-long, multiple-feature branches.
+     - The larger the PR, the harder it is to review, the more likely it will need to be rebased, and the harder it is to rebase it.
+     - If a PR changes thousands of lines of code, please consider splitting it into smaller PRs.
+   - Developers should aim to have their code merged at least once every week.
+   - Multiple small fixes or refactorings can be grouped together in one PR, but each independent fix or refactoring should be a unique commit.
+* All code must be reviewed before it is merged to the master branch.
```
Comment (eu-smirnov) [inline CONTRIBUTING.md]: I'd probably add something about commits in PR, like:
 - Make separate commits for logically separate changes within the PR.
 - Each commit should be buildable and ideally passing the tests, to simplify git bisect.
 - Short description should not exceed 50 chars.

Comment (patrickvacek) [reply CONTRIBUTING.md]: > Make separate commits for logically separate changes within the PR.

Agreed.

> Each commit should be buildable and ideally passing the tests, to simplify git bisect.

Generally agreed, but I'd consider an exception for the later line ("Bugs should be reproduced with a failing test case before they are resolved.").

> Short description should not exceed 50 chars.

Does it really have to be 50? That's so short! I'd prefer something around 70-80, which is closer to what we do in practice.

Comment (eu-smirnov) [reply CONTRIBUTING.md]: 50 is what usually recommended, e.g. https://gerrit.googlesource.com/plugins/commit-message-length-validator/+doc/master/src/main/resources/Documentation/config.md
But it can definitely break after 72 chars, so maybe we can put this limit here.

TBH, I'm not entirely sure about mentioning tests in p.2. I doubt that anybody will really check that.

Comment (patrickvacek) [reply CONTRIBUTING.md]: Updated (and rebased). There is some redundancy in the two lines about commits, but that's the only (minor) concern I have. I made the comment about tests passing for each commit more explicitly an ideal than a requirement, but I think it's fine to keep that as an aspiration.
